.github/workflows/docker-publish.yml

```diff
@@ -0,0 +1,198 @@
+
+      - name: Use cache
+        uses: actions/cache@v2
+        with:
+          path: /tmp/.build-cache
+          key: ${{ runner.os }}-docker-${{ github.sha }}
+
+      - name: Build Docker image
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          push: false
+          tags: ${{ env.TEST_TAG }}
+          file: docker/Dockerfile
+          cache-from: type=local,src=/tmp/.build-cache
+          load: true
+
+      - name: Run test recipes on host
+        run: |
+          docker-compose -f docker/recipes.test.yml \
+            up --exit-code-from=sut
+        env:
+          IMAGE: ${{ env.TEST_TAG }}
+
+  uml-test:
+    name: Run UML tests
+    needs: [build]
+    runs-on: ubuntu-latest
+    steps:
+      - name: Repository checkout
+        uses: actions/checkout@v2
+
+      - name: Setup Docker buildx
+        uses: docker/setup-buildx-action@v1
+
+      - name: Use cache
+        uses: actions/cache@v2
+        with:
+          path: /tmp/.build-cache
+          key: ${{ runner.os }}-docker-${{ github.sha }}
+
+      - name: Build Docker image
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          push: false
+          tags: ${{ env.TEST_TAG }}
+          file: docker/Dockerfile
+          cache-from: type=local,src=/tmp/.build-cache
+          load: true
+
+      - name: Run test recipes using UML backend
+        run: |
+          docker-compose -f docker/recipes-test-uml.yml \
+            up --exit-code-from=sut
+        env:
+          IMAGE: ${{ env.TEST_TAG }}
+
+  publish:
+    name: Publish
+    needs:
+      - unit-test
+      - recipe-test
+      - uml-test
+    if: github.event_name != 'pull_request'
+    runs-on: ubuntu-latest
+    permissions:
+      contents: read
+      packages: write
+
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v2
+
+      - name: Extract Docker metadata
+        id: meta
+        uses: docker/metadata-action@v3
+        with:
+          images: ${{ env.DOCKERHUB_TAG }},${{ env.GITHUB_TAG }}
+
+      - name: Login to GitHub Container Registry
+        uses: docker/login-action@v1
+        with:
+          registry: ghcr.io
+          username: ${{ github.repository_owner }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Login to DockerHub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_PASSWORD }}
+
+      - name: Setup Docker buildx
+        uses: docker/setup-buildx-action@v1
+
+      - name: Use cache
+        uses: actions/cache@v2
+        with:
+          path: /tmp/.build-cache
+          key: ${{ runner.os }}-docker-${{ github.sha }}
+
+      - name: Build and push Docker image
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          push: true
+          tags: ${{ steps.meta.outputs.tags }}
```

Comment:
since it says tags; i guess it pushes to an array of tags ?

Reply:
The meta step builds up the tags here; we pass it two bases `ghcr.io/go-debos/debos` and `godebos/debos`. It then appends something useful - ":master" or e.g. ":v1.0.1" to both. Those then become the two targets for the image.

Reply:
it should append `:latest`, not `:master` ?

Reply:
I can look into that. It's the default for the Github setup (which I think is master here). Will investigate.

Reply:
The default is to tag pushes to master just as :master, and to only add :latest when a version is tagged:

> ### Latest tag
> latest tag is handled through the flavor input. It will be generated by default (auto mode) for:
>  -  type=ref,event=tag
>  - type=semver,pattern=...
>  - type=match,pattern=...

The default table of examples (which we're using here) looks like it has semver setup, giving:

> Event | Ref | Docker Tags
> -- | -- | --
> push | refs/heads/master | master
> push tag | refs/tags/v1.2.3 | v1.2.3, latest
> push tag | refs/tags/v2.0.8-beta.67 | v2.0.8-beta.67, latest

I could change it to tag master as latest as well, if that's desired. On the face of it, only tagging releases feels like quite a sensible default, but happy to change this @obbardc?

Reply:
Hm, we don't actually do many releases. Currently in Dockerhub the `latest` tag points to the last build `HEAD`.

Reply:
Perhaps we should do a nightly scheduled build of master as a separate workflow? I'd be happy to do that if that sounds useful? It avoids some of the intricacies of keeping `:latest` up to date when it builds from different triggers.

Reply:
I think we just need to make sure that `latest` tag points to the last container built from `master` ;-)
Wasting actions time as a scheduled run is... meh

Reply:
I've done this, but the only way to avoid hard-coding master (given the lack of variable support, and the general state of conditionals) was to simply distinguish any push. I think this is safe in this repo, because there's only one branch. It isn't particularly good in forks though, that might have more branches, but I'm not sure how to resolve this. I could filter every step with the test against the environment, but that will lead to all four publish actions running, which is harder to follow (presently, two of the four run for any given workflow run, with the two that aren't appropriate marked as disabled).